handel/src/aggregation.rs

```diff
@@ -326,100 +332,190 @@ impl<
+        let mut best_aggregate = if let Some(verification_future) = &mut self.current_verification {
+            if let Poll::Ready((result, todo)) = verification_future.poll_unpin(cx) {
+                // If a result is produced, unset the future such a new one can take its place.
+                self.current_verification = None;
+                if result.is_ok() {
+                    // If the todo was successfully verified, apply it and return the new best aggregate
+                    Some(self.apply_todo(todo))
+                } else {
+                    // If the Todo failed to verify there is nothing to be done, and also no new best aggregate.
+                    log::debug!(?result, ?todo, "Verification of Todo Item failed.");
+                    None
+                }
+            } else {
+                // The future has not yet resolved. There is no new best aggregate.
+                None
+            }
+        } else {
+            // There is no future to poll and thus no new todo to process. There is no new best aggregate.
+            None
+        };
+
+        // Check and see if a new todo should be verified.
+        // Only start a new verification task if there is not already one and also only if the poll does not have produced a value yet.
+        // This is necessary as the verification future could resolve immediately producing a second item for the stream.
+        // As the new best aggregate will be returned this stream will be polled again creating the future in the next poll
+        if self.current_verification.is_none() && best_aggregate.is_none() {
+            // Get the next best todo.
+            while let Poll::Ready(Some(todo)) = self.todos.poll_next_unpin(cx) {
+                // Start the verification. This will also poll and thus may immediately return a result.
+                if let Some((result, todo)) = self.start_todo_verification(todo, cx) {
+                    // If the future returned immediately it needs to be handled.
+                    if result.is_ok() {
+                        // If the todo verified, apply it and return the new best aggregate
+                        let new_aggregate = self.apply_todo(todo);
+
+                        best_aggregate = Some(new_aggregate);
+                        break;
+                    } else {
+                        log::debug!(?result, ?todo, "Verification of Todo Item failed.");
                     }
-                },
+                } else {
+                    break;
+                }
             }
         }
-    }
 
-    fn into_inner(self) -> (LevelUpdateStream<P, TId>, LevelUpdateSender<N>) {
-        (self.todos.into_stream(), self.sender)
+        // Poll sender to until it is pending sending as many messages as possible.
+        while let Poll::Ready(t) = self.sender.poll_next_unpin(cx) {
+            t.expect("Sender should never return None");
+        }
```

Comment:
It seems that the `LevelUpdateSender` can never return `Poll::Ready`

```suggestion
        assert!(self.sender.poll_next_unpin(cx).is_pending());
```

Since that is the case, maybe it shouldn't implement `Stream`, but rather `Future`.

Reply:
Good find! I made it a Future instead. I also added the asserts.